Add SAML Response, Logout Request and Logout Response validations. Add also tests and documentation

pitbulk opened the pull request:
This PR is a meta-PR:
* Add SAML Response, Logout Request and Logout Response validations. (based on this old PR: https://github.com/onelogin/ruby-saml/pull/115)
* Add tests (high coverage, use of Minitest)
* Add HTTP-Redirect Binding Validation Sign (https://github.com/onelogin/ruby-saml/commit/42a7bbd04b730bf248e33591aa0139da2098b4d5)
* Add documentation to the README (https://github.com/onelogin/ruby-saml/issues/187, https://github.com/onelogin/ruby-saml/issues/186) and in the code (YARD style)
* Fix XMLSecurity class: (SHA1 vs RSA_SHA1 issue)
* Ensure IdP certificate is formatted properly (Related to https://github.com/onelogin/ruby-saml/pull/176)

pitbulk commented:
@luisvm @Lordnibbler @daniel-g Can you review this PR? 
(You can pull the branch and update changes by yourself)

Lordnibbler commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -67,46 +98,200 @@ def status_code
+      # If fails, the attribute errors will contains the reason for the invalidation.
+      # @param [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the logout response is invalid or not)
+      # @param [String|nil] request_id The ID of the Logout Request sent by this SP to the IdP (if was sent any)
+      # @return [Boolean|ValidationError] True if the Logout Response is valid, otherwise
```
```ruby
# Validates the Logout Response (calls several validation methods)
# If fails, the attribute errors will contains the reason for the invalidation.
# @param [Boolean] soft Enable or Disable the soft mode in order to raise exceptions if logout response is invalid
# @param [String|nil] request_id The ID of the Logout Request sent by this SP to the IdP (if was sent any)
# @return [Boolean|ValidationError] true if the Logout Response is valid, otherwise false if soft == true
# @raise [ValidationError] if soft == false
```

pitbulk replied:
Ok I will see what YARD generates and fix the documentation to be compatible with YARD.

Lordnibbler commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -6,19 +6,30 @@
       DSIG      = "http://www.w3.org/2000/09/xmldsig#"
 
-      # TODO: This should probably be ctor initialized too... WDYT?
+      # TODO: Settings should probably be initialized too... WDYT?
```
maybe address this todo, and initialize `@settings` in the `initialize` method

pitbulk replied:
This means documentation changes and compatibility issues.
I postponed this knowing that we may change that soon.

luisvm replied:
`settings` could be loaded in the `initialize` with no compatibility harm since the constructor accepts optional hashes, you should just check for `options[:settings]` and if it's not nil, assign `@settings` - otherwise it would be working as before

phlipper replied:
:+1: @luisvm

Lordnibbler commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -27,19 +38,32 @@ def initialize(response, options = {})
         @document = XMLSecurity::SignedDocument.new(@response, @errors)
       end
 
+      # An aux function to validate the SAML Response with the default values (soft = true)
```
no reason to have these comments `An aux function...`.  Keep documentation short and sweet

Lordnibbler commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -27,19 +38,32 @@ def initialize(response, options = {})
 
-      def validate!
-        validate(false)
+      # Another aux function to validate the SAML Response (soft = false)
```
`validate the SAML Response (soft = false)`

Lordnibbler commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -27,19 +38,32 @@ def initialize(response, options = {})
+      # After execute a validation process, if fails this method returns the causes
+      # @return [Array] Empty Array if no errors, or an Array with the causes
+      #
       def errors
```
unnecessary method, simply call `self.errors`, `@errors`, or `errors`

Lordnibbler commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -27,19 +38,32 @@ def initialize(response, options = {})
         @errors
       end
 
-      # The value of the user identifier as designated by the initialization request response
```
Simplify the documentation:

```ruby
# @return [String] the NameID provided by the SAML response from the IdP.
```

Lordnibbler commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -110,52 +158,259 @@ def status_message
+        end
+      end
+
+      # Gets the Destionation attribute from the SAML Response.
```
s/Destionation/Destination

Lordnibbler commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -110,52 +158,259 @@ def status_message
+      end
+
+      # Gets the Destionation attribute from the SAML Response.
+      # @return [String|nil] The Destionation value if exists.
```
```ruby
# @return [String|nil] destination attribute from the SAML Response
```

Lordnibbler commented:
As a general comment, let's add Yard as a development dependency and generate the docs.  I think there are a number of cases where the documentation does not follow the YARD syntax.

Lordnibbler commented on test/test_helper.rb:
```diff
@@ -60,6 +71,102 @@ def response_document_7
     @response_document7 ||= Base64.encode64(File.read(File.join(File.dirname(__FILE__), 'responses', 'response_no_cert_and_encrypted_attrs.xml')))
   end
 
+  def response_document_8
```
i dont see this being used?

Lordnibbler commented on test/test_helper.rb:
```diff
@@ -60,6 +71,102 @@ def response_document_7
+    @response_document8 ||= File.read(File.join(File.dirname(__FILE__), 'responses', 'response8.xml.base64'))
+  end
+
+  def response_document_9
```
i dont see this being used?

Lordnibbler commented on test/utils_test.rb:
```diff
@@ -0,0 +1,171 @@
+
+  describe "#format_cert" do
+    it "return empty string when the cert is an empty string" do
+      cert = ""
```
for setup, I'd recommend taking advantage of minitest's syntax sugar

```ruby
describe '#format_cert' do
  let(:cert) { '' }
 
  it 'returns empty string when the cert is an empty string' do
    # test logic
  end 
```

phlipper replied:
:+1:

Lordnibbler commented on test/utils_test.rb:
```diff
@@ -0,0 +1,171 @@
+      assert formatted_cert.include? "-----END CERTIFICATE-----"
+    end
+
+    it "return formatted cert with head from cert without head" do
```
you can do all of this cert checking logic with Ruby's `OpenSSL` module.  ie. `OpenSSL::X509::Certificate.new(cert_string)`

http://ruby-doc.org/stdlib-2.0/libdoc/openssl/rdoc/OpenSSL.html

pitbulk replied:
Yes I will replace this method and merge here the https://github.com/onelogin/ruby-saml/pull/176

Lordnibbler replied:
pretty much anywhere you are parsing certs should be done with `OpenSSL` module IMO

pitbulk replied:
@Lordnibbler I can see at OpenSSL any method to format the cert/private_key.

Imagine those scenarios:

A user copy a cert from a website where the public cert file is printed, but there are extra new lines, when he paste the cert in the ruby-saml setting file, the 
```OpenSSL::X509::Certificate.new(cert_string)```
command will fail. If we previously call the format_cert method, the user will not experience any problem.

I added here a test so you can review what I said:
https://github.com/onelogin/ruby-saml/commit/da8c9ec1679cbe66ec6be5411b48a9faa63f9a28

Lordnibbler commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -5,62 +5,269 @@
+      # After execute a validation process, if fails this method returns the causes
+      # @return [Array] Empty Array if no errors, or an Array with the causes
+      #
+      def errors
```
not necessary

Lordnibbler commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -67,46 +98,200 @@ def status_code
+      # After execute a validation process, if fails this method returns the causes
+      # @return [Array] Empty Array if no errors, or an Array with the causes
+      #
+      def errors
```
not necessary

Lordnibbler commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -87,22 +119,38 @@ def attributes
         @status_code ||= begin
           node = REXML::XPath.first(document, "/p:Response/p:Status/p:StatusCode", { "p" => PROTOCOL, "a" => ASSERTION })
-          node.attributes["Value"] == "urn:oasis:names:tc:SAML:2.0:status:Success"
+          node.attributes["Value"] if node and not node.attributes.nil?
```
`if node && node.attributes`

luisvm commented on lib/onelogin/ruby-saml/attributes.rb:
```diff
@@ -1,91 +1,110 @@
       # By default Attributes#[] is backwards compatible and
       # returns only the first value for the attribute
-      # Setting this to `false` returns all values for an attribute
+      # Setting this to `false` returns all values for an attribute      
```
nitpick: let's be careful with adding trailing whitespaces, might want to add a hook to your editor to remove these

Lordnibbler replied:
+9000

luisvm commented on lib/onelogin/ruby-saml/utils.rb:
```diff
@@ -1,44 +1,83 @@
+      # @param sig_alg     [String] The SigAlg parameter                  
+      # @return [String] The Query String 
+      #
+      def self.build_query(type, data, relay_state, sig_alg)
```
If we're accepting this many parameters, my 5 cents would be to use an options hash so when we call `Utils.build_query` it's readable to which params we're sending, besides, the order of the params wouldn't matter.

daniel-g replied:
+1

luisvm commented on lib/onelogin/ruby-saml/utils.rb:
```diff
@@ -1,44 +1,83 @@
+      # @param query_string [String] The SigAlg parameter
+      # @return [Boolean] True if the Signature is valid, False otherwise
+      #
+      def self.verify_signature(cert, sig_alg, signature, query_string)
```
same comment as above

daniel-g commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +103,202 @@ def status_code
+      #
+      def current_url
+        @current_url ||= begin
+          unless self.settings.nil? or self.settings.single_logout_service_url.nil?
```
use || instead of or. or is more for function executions, not for conditionals

luisvm replied:
+1

luisvm replied:
I'd also recommend using `if` instead of `unless` here to make this more readable:
```ruby
if settings && settings.single_logout_service_url
  ...
end
```
I'd also avoid using `self` unless there's a method that could be confused with a passed variable with the same name, see how on the previous methods we don't use `self.document` to get the values for that accessor

daniel-g commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +103,202 @@ def status_code
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_success_status(soft = true)
+        if success?
```
you can use an eager return:
```
return true if success?
...the else code here...
```

daniel-g commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +103,202 @@ def status_code
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_destination(soft = true)
+        return true if destination.nil? or destination.empty? or settings.single_logout_service_url.nil? or settings.single_logout_service_url.empty?
```
use || here

luisvm replied:
agree, use `or` like in PHP:
```php
validate() or die('fail');
```

daniel-g commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +103,202 @@ def status_code
+        return true if get_params.nil? || get_params['Signature'].nil? || self.settings.nil? || self.settings.get_idp_cert.nil?
+        
+        require "byebug"
+        debugger
```
this shouldn't be here, I guess.

luisvm replied:
let's remove these two lines

daniel-g commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +103,202 @@ def status_code
+        require "byebug"
+        debugger
+
+        query_string = OneLogin::RubySaml::Utils.build_query('SAMLResponse',
```
nit pick:
```
method(
  'param1',
  'param2',
  'param3',
)
```
makes it more readable

daniel-g commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -29,19 +40,25 @@ def initialize(response, options = {})
+      # @return [Boolean] TRUE if the SAML Response is valid
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate!(soft=false, request_id = nil)
```
sometimes this kind of parameters makes it a lot harder to know what is going on exactly. Instead of method(true, 1234), use method(:soft => true, :request_id => 1234)

luisvm replied:
I would think that by using `validate!` I would expect it to raise exceptions, so there wouldn't be the need to accept `soft` and just use it as `validate(false, request_id)` here - but I'm not entirely sure about this convention... any word @daniel-g ?

daniel-g replied:
raising exceptions depend on the team, but most of the time I've seen that's the convention.

phlipper replied:
Using _bang!_ methods conventionally indicates a "destructive" behavior, and it is very reasonable to expect an exception to be raised here. I'd ideally like to see the the validation code get rid of all of this `soft` noise and simply call the error routine, which can itself either `raise`, or potentially just return the error messages based on settings.

Lordnibbler replied:
I agree with @phene here, but this might be out of the scope for this specific PR

luisvm replied:
thanks @phlipper - I also agree with your comment

daniel-g commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,327 @@ def status_message
+          audiences = []
+          nodes = xpath_from_signed_assertion('/a:Conditions/a:AudienceRestriction/a:Audience')
+          nodes.each do |node|
+            unless node.nil? or node.text.empty?
```
use ||

daniel-g replied:
aso, for this kind of negatives I'd prefer the positive conditional:
if !node.nil? and !node.text.empty?

luisvm replied:
`if node && node.text` would suffice imo

daniel-g commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,327 @@ def status_message
+      def validate(soft = true, request_id = nil)
+        @errors = []
         validate_response_state(soft) &&
+        validate_id                   &&
```
lots of validations... have you considered using a gem for that?

phlipper replied:
Minor nitpick, but all of the whitespace after the validation and before the `&&` makes it hard to understand that this is one continuous idea ... remove the whitespace, and indent the overflow lines:

```ruby
validate_response_state(soft) &&
  validate_id &&
  validate_other &&
  #...
```

daniel-g commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -228,20 +547,110 @@ def validate_conditions(soft = true)
+        
+        now = Time.now.utc
+        subject_confirmation_nodes.each do |subject_confirmation|
+          if subject_confirmation.attributes.include? "Method" and subject_confirmation.attributes['Method'] != 'urn:oasis:names:tc:SAML:2.0:cm:bearer'
```
holy if party!! any chance to refactor?

pitbulk replied:
I really don't know how refactor that

I have the same scheme in PHP:
https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Response.php#L205

but I think is easy to read...

luisvm replied:
```ruby
next if subject_confirmation.attributes['Method'] != 'urn:oasis:names:tc:SAML:2.0:cm:bearer'

phlipper replied:
:+1: for @luisvm's comment. Also, I see that @Lordnibbler already linked you to Avdi's post about _and/or in Ruby_, and there are a few spots left _(including here)_ that need to be updated.

Lordnibbler commented on Rakefile:
```diff
@@ -4,8 +4,8 @@ require 'rake'
 Rake::TestTask.new(:test) do |test|
-  test.libs << 'lib' << 'test'
-  test.pattern = 'test/**/*_test.rb'
+  test.libs << "lib" << "test"
```
stick to `'` unless you are using string interpolation as a rule of thumb

daniel-g commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -7,60 +7,268 @@
+        #
+        def current_url
+          @current_url ||= begin
+            unless self.settings.nil? or self.settings.single_logout_service_url.nil?
```
use ||

daniel-g commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -7,60 +7,268 @@
+        #
+        def validate(soft = true, get_params = nil)
+          @errors = []
+          validate_request_state(soft) &&
```
some of them are soft and some are not?

phlipper replied:
Please fix this inconsistency.

daniel-g commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -7,60 +7,268 @@
+        # @raise [ValidationError] if soft == false and validation fails
+        #
+        def validate_destination(soft = true)
+          return true if destination.nil? or destination.empty? or settings.single_logout_service_url.nil? or settings.single_logout_service_url.empty?
```
switch to ||

daniel-g commented on test/logoutrequest_test.rb:
```diff
@@ -133,22 +121,24 @@ class RequestTest < Minitest::Test
-        @settings.private_key = ruby_saml_key_text
-        @cert = OpenSSL::X509::Certificate.new(ruby_saml_cert_text)
+
+      let(:settings3) { OneLogin::RubySaml::Settings.new }
```
settings3 doesn't seem very descriptive. What kind of settings is this?

phlipper replied:
@pitbulk I can't disagree more ... descriptive names make the code **much** easier to read. There are better ways to deal with line length than making hard-to-understand variable names.

Lordnibbler replied:
@pitbulk also agree, lets choose better variable names here

luisvm commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,327 @@ def status_message
+      def issuers
+        @issuers ||= begin
+          issuers = []
           node = REXML::XPath.first(document, "/p:Response/a:Issuer", { "p" => PROTOCOL, "a" => ASSERTION })
```
try this with the XPath `/p:Response/a:Issuer | /p:Response/a:Assertion/a:Issuer` so you can get rid of line 184-191 and implement:
```ruby
nodes.each do |node|
  issuers << node.text
end
```

luisvm commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,327 @@ def status_message
+        assertions = REXML::XPath.match(document, "//a:Assertion", { "a" => ASSERTION })
+        encrypted_assertions = REXML::XPath.match(document, "//a:EncryptedAssertion", { "a" => ASSERTION })
+
+        unless assertions.length + encrypted_assertions.length == 1
```
just a suggestion (personal preference) when using `length` or `size`:
```
Whether you should use length or size is mostly a matter of personal preference. Personally I use size for collections (hashes, arrays, etc) and length for strings, since for me objects like hashes and stacks don’t have a length, but a size (defined in terms of the elements they contain)
```

http://batsov.com/articles/2014/02/17/the-elements-of-style-in-ruby-number-13-length-vs-size-vs-count/

luisvm commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -228,20 +547,110 @@ def validate_conditions(soft = true)
+          confirmation_data_node = REXML::XPath.first(subject_confirmation, 'a:SubjectConfirmationData', { "a" => ASSERTION })
+
+          if not confirmation_data_node
+            next
```
`next unless confirmation_data_node` so that you don't use `else` below

phlipper replied:
:+1: @luisvm

luisvm commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -228,20 +547,110 @@ def validate_conditions(soft = true)
+          if not confirmation_data_node
+            next
+          else
+            if confirmation_data_node.attributes.include? "InResponseTo" and confirmation_data_node.attributes['InResponseTo'] != in_response_to
```
if the `confirmation_data_node.attributes` hash doesn't include `InResponseTo` key, doing `confirmation_data_node.attributes['InResponseTo']` will return `nil` so you can skip `include?` steps here and below

luisvm commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -7,60 +7,268 @@
       def name_id
         @name_id ||= begin
-          node = REXML::XPath.first(document, "/p:LogoutRequest/a:NameID", { "p" => PROTOCOL, "a" => ASSERTION })
+          node = REXML::XPath.first(self.document, "/p:LogoutRequest/a:NameID", { "p" => PROTOCOL, "a" => ASSERTION })
```
no need to use `self`

luisvm commented on lib/onelogin/ruby-saml/utils.rb:
```diff
@@ -1,44 +1,83 @@
-          cert = cert.gsub('-----END CERTIFICATE-----', '')
-          cert = cert.gsub(' ', '')
-
+        unless cert.empty?
```
was there a problem using `if cert` ?

pitbulk replied:
Can I use?
```
if cert && cert != ""
```

luisvm replied:
sure! that sounds good +1 `empty?` also works but you just need to check for `cert` first:
`if cert && !cert.empty?`

luisvm commented on lib/onelogin/ruby-saml/utils.rb:
```diff
@@ -1,44 +1,83 @@
-        key = key.delete("\n").delete("\r").delete("\x0D")
-        if key
+        unless key.empty?
+          key = key.delete("\n").delete("\r").delete("\x0D")
```
```ruby
key.delete!("\n\r\x0D")
```

pitbulk commented:
@luisvm, @Lordnibbler can we merge this PR now?

phlipper commented:
This is an overwhelmingly large PR, and there are several very different bits of context all crammed in here, making a comprehensive review challenging. This should really be 3 separate PRs _(one for adding doc comments, one for adding tests, and one for adding the response validations)_. 

This also contains partial content from other PRs _(like #176, for example)_, but not the full changes, which makes things more challenging to realign after this gets merged. 

Please at least address the remaining comments in this PR, as there are several decent suggestions which will help ease the pain of future maintenance a bit. 

It would also be good to at least rebase away commits like _"Fix typo"_ and _Merge_ commits, and make commit messages generally more useful than simply _"code clean"_. Commit messages should explain _why_ a change was made, not simply _what_ was changed, which is what the diff is for.

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate(soft = true, request_id = nil, get_params = nil)
+        @errors = []
```
This is re-initialized here from Line 31

pitbulk replied:
@supernullset , Yes, notice that if we call the validation process twice, we don't want to have duplicate message errors.

supernullset replied:
That makes some sense from an external standpoint, but that does speak that `@errors` should probably be memoized, Initializing the ivar in multiple places can lead to unforeseen issues in debug/development.

Bringing `@errors` into a method like

```ruby
def errors
  @errors ||= []
end

# with another method to bust that memoization when you need it

def reset_errors!
  @errors = []
end
```

would be safer.

Lordnibbler replied:
+1

phlipper commented on lib/onelogin/ruby-saml/authrequest.rb:
```diff
@@ -46,11 +64,13 @@ def create_params(settings, params={})
+            :relay_state => relay_state,
+            :sig_alg     => params['SigAlg']
+          )
+          signature           = settings.get_sp_key().sign(XMLSecurity::BaseDocument.new.algorithm(settings.security[:signature_method]).new, url_string)
```
Unnecessary parens here. Also, remove the extra whitespace before `=`, and assign the xml security algorithm to a variable on the line above and use it here _(vs. inlining the call)_ for readability.

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+      # @return [Boolean] True if the Logout Response is valid, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate(soft = true, request_id = nil, get_params = nil)
```
I think that the logic of this could be significantly cleaner.

If there was a boolean `@valid`(or another name with the same intent) instance variable on this class, you could drop all of the `&&` and simply mutate that variable to be returned on line 146. All of the validation methods are mutating ivars anyway, adding another would be a minimal change.

supernullset replied:
Also, looking at this, `soft` is used enough in this validation step that its probably worth making it an ivar which lives in the validation life-cycle. Passing around the same boolean in 7 methods seems superfluous.

Lordnibbler replied:
+1

phlipper commented on lib/onelogin/ruby-saml/logoutrequest.rb:
```diff
@@ -44,11 +62,13 @@ def create_params(settings, params={})
+            :relay_state => relay_state,
+            :sig_alg     => params['SigAlg']
+          )
+          signature           = settings.get_sp_key().sign(XMLSecurity::BaseDocument.new.algorithm(settings.security[:signature_method]).new, url_string)
```
Same comments about parens/whitespace/algorithm as above.

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
-        if response.empty?
-          return soft ? false : validation_error("Blank response")
+        if response.nil? or response.empty?
+          error_msg = "Blank Logout Response"
```
the 

```ruby
error_msg = ...
@errors << error_msg
<ternary>
```
Construct is reused enough here that pulling it out into its own method would be worthwhile.

```ruby
def append_error(soft, error_msg)
  @errors << error_msg
   return soft ? false : validation_error(error_msg)
end
```

pitbulk replied:
Ok, noted

phlipper commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -3,50 +3,61 @@
 
+      # @return [Boolean] TRUE if the SAML Response is valid
+      #
+      def validate!(soft=false, request_id = nil, get_params = nil)
```
What is the point of this method if it does nothing but wrap `validate` without the bang? This just makes more noise.

pitbulk replied:
The Response method got it, so I added this to the logoutresponse too (compatibility).
Do you think we may remove it?

phlipper replied:
Does it need to be here? If so, the usage should be consistent across the library. Perhaps simply rename `validate` to `validate!` _(though note my prior comments about expectation of bang methods)_. If not, please just remove it.

phlipper commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -54,14 +65,38 @@ def in_response_to
+      # 
+      def success?(soft = true)
+        unless status_code == "urn:oasis:names:tc:SAML:2.0:status:Success"
+          return soft ? false : validation_error("Bad status code. Expected <urn:oasis:names:tc:SAML:2.0:status:Success>, but was: <#@status_code> ")
```
Please use proper interpolation here, the `#@status_code` syntax is asking for trouble, and will be incompatible with future versions of Ruby.

Lordnibbler replied:
```ruby
"Bad status code. Expected <urn:oasis:names:tc:SAML:2.0:status:Success>, but was: #{@status_code} "
```

pitbulk commented:
@phlipper I updated the description of the PR.
I apologize for how this PR became a nightmare. I added the SAML validations that carried a lot of code changes, after that I wanted to improve the test coverage and the documentation of the code.. and meanwhile I discovered several bugs that I wanted to include.
Inn order to avoid rebase coding a lot of minor PRs, I tried to include them here (some of them got unwanted parts of code or where without documentation so I improved them here).

The main problem is that the merged process of this PR took forever and the project was so active meanwhile. Notice that the PR was released on 3 Feb and almost 2 months past.

I will do my best and I promise no generate a PR like this anymore. :)

phlipper commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+
+      # Gets the expected current_url
+      # (Right now we read this url from the Sinle Logout Service of the Settings)
+      # TODO: Calculate the real current_url and use it.
```
Why not handle this _TODO_ now, since this is being introduced here?

pitbulk replied:
I think this method not gonna be trivial because depends on what ruby web framework be used.
We need to think in a global solution compatible with any framework and don't want to delay more this merge.

phlipper replied:
@pitbulk why does this need to depend on any framework? This library should be completely self-contained, with extension points if/as needed. This library makes a lot of assumptions about how it _might_ be used, which adds a lot of mental, as well as code, overhead.

pitbulk replied:
@phlipper do you have any general get_current_url method?

In python toolkit we use https://github.com/onelogin/python-saml/blob/master/src/onelogin/saml2/utils.py#L320 that depends on the request_data and each framework prepare this value:
https://github.com/onelogin/python-saml/blob/master/demo-django/demo/views.py#L17

In PHP is easier since the values to build the current_url are at the global $_SERVER var: https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Utils.php#L340

A first approach is to check that the 'Destination' value match the ACS URL, later I want to validate that the real destination match the 'Destination' value.

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+            @errors << "Invalid Logout Response. Not match the saml-schema-protocol-2.0.xsd"
+          end
+          valid
+        rescue OneLogin::RubySaml::ValidationError => e
```
While I realize that this pattern may have been here before this PR was opened, IME using exceptions in a validation context leads to bad JuJu. An error state is not necessarily exceptional, it is only exceptional if you have *not* accounted for the case.

This may be a comment outside the context of this PR.

phlipper commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+      # @return [String] The current url
+      #
+      def current_url
+        @current_url ||= begin
```
Does this really need to be memoized? Also this can be simplified to simply:

```ruby
@current_url ||= settings && settings.single_logout_service_url
```

pitbulk replied:
The code will be replaced by the calculate current url method.

phlipper commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
       def valid_state?(soft = true)
-        if response.empty?
-          return soft ? false : validation_error("Blank response")
+        if response.nil? or response.empty?
```
Use `||` not `or` here

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+          error_msg +=  ' -> ' + status_message
+        end
+
+        @errors << error_msg
```
Same pattern as pointed out above; this should hop into a method.

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+        return true if request_id.nil?
+
+        unless request_id == in_response_to
+          error_msg = "Logout Response does not match the request ID, expected: <#{request_id}>, but was: <#{in_response_to}>"
```
Same pattern as above, this should hop into a method

phlipper commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+          end
+          valid
+        rescue OneLogin::RubySaml::ValidationError => e
+          @errors << e.message
```
What does adding to the `@errors` array do here if you're just re-raising on the next line?

pitbulk replied:
I simple wanted the cause to be available at errors.

phlipper replied:
@pitbulk I still don't understand. You don't have access to `@errors` once the exception is raised, execution just stops. Perhaps the errors should be added to the exception?

pitbulk replied:
If someone that use the library put the validation code inside a rescue, will have the cause on the @errors. If still not makes sense for you, I can just remove it.

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
-        unless self.options[:matches_request_id] == in_response_to
-          return soft ? false : validation_error("Response does not match the request ID, expected: <#{self.options[:matches_request_id]}>, but was: <#{in_response_to}>")
+        unless destination == current_url
+          error_msg = "The Logout Response was received at #{destination} instead of #{current_url}"
```
Same pattern as above, this should hop into a method

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
-        unless URI.parse(self.issuer) == URI.parse(self.settings.idp_entity_id)
-          return soft ? false : validation_error("Doesn't match the issuer, expected: <#{self.settings.issuer}>, but was: <#{issuer}>")
+        unless URI.parse(issuer) == URI.parse(settings.idp_entity_id)
+          error_msg = "Doesn't match the issuer, expected: <#{settings.idp_entity_id}>, but was: <#{issuer}>"
```
Same pattern as above, this should hop into a method

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+      #
       def valid_issuer?(soft = true)
-        return true if self.settings.idp_entity_id.nil? or self.issuer.nil?
+        return true if settings.idp_entity_id.nil? or issuer.nil?
```
IMO this deserves parens around the `or` condition

phlipper commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+        end
+
+        @errors << error_msg
+        soft ? false : validation_error(error_msg)
```
Most places in the changeset use an explicit return for lines like these. While I prefer the implicit return style, I'd rather see consistent usage across this library.

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+      # @raise [ValidationError] if soft == false and validation fails
+      #      
+      def validate_signature(soft = true, get_params = nil)
+        return true if get_params.nil? || get_params['Signature'].nil? || settings.nil? || settings.get_idp_cert.nil?
```
the `or` condition here should be moved into a boolean method with a descriptive name.

supernullset commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+        )
+
+        unless valid
+          error_msg = "Invalid Signature on Logout Response"
```
Same pattern as above, this should hop into a method

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+          issuers = []
+          nodes = REXML::XPath.match(document, "/p:Response/a:Issuer | /p:Response/a:Assertion/a:Issuer", { "p" => PROTOCOL, "a" => ASSERTION })
+          nodes.each do |node|
+            issuers << node.text if not node.text.nil?
```
Use `!` vs. `not` here. Also, simply `issuers << node.text if node.text` should work here.

luisvm replied:
+1 agree

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
-        valid_saml?(document, soft)      &&
+      # Gets the expected current_url
+      # (Right now we read this url from the Assertion Consumer Service of the Settings)
+      # TODO: Calculate the real current_url and use it.
```
This is either a copy-pasta comment, or this should really be addressed in this changeset.

Lordnibbler replied:
+1

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+      # @return [String] The current url
+      #
+      def current_url
+        @current_url ||= begin
```
This memoization isn't necessary.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response contains an ID, otherwise returns False
+      #
+      def validate_id()
```
Remove these empty parens

supernullset commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+      # @return [Boolean] True if the SAML Response is valid, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate(soft = true, request_id = nil)
```
I added some comments about validation in lib/onelogin/ruby-saml/logoutresponse.rb I think that most if not all of them apply here.

There is enough overlap in the conceptual ideas across validation in these files that I question the repetition. I think even if nothing is changed with respect to my comments(no biggie) you should *definetly* pull out the shared validation logic into another file. By moving over the shared logic you will be able to more appropriately test, and debug issues.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response is 2.0, otherwise returns False
+      #
+      def validate_version()
```
Remove these empty parens

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response contains one unique Assertion, otherwise False
+      #
+      def validate_num_assertion()
```
Remove these empty parens

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
-          soft ? false : validation_error(status_message)
+          error_msg = 'The status code of the Response was not Success'
+          unless status_code.nil?
+            printable_code = status_code.split(':').pop
```
Using `.last` vs. `.pop` here would be more intention-revealing.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+          end
+
+          @errors << error_msg
+          soft ? false : validation_error(error_msg)
```
Inconsistent missing `return`.

phlipper commented on lib/onelogin/ruby-saml/logoutresponse.rb:
```diff
@@ -69,46 +104,211 @@ def status_code
+
+        error_msg = 'The status code of the Logout Response was not Success'
+        unless status_code.nil?
+          printable_code = status_code.split(':').pop
```
Using .last vs. .pop here would be more intention-revealing.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+      # @return [Boolean] True if the SAML Response contains a Success code, otherwise False if soft == false
+      # @raise [ValidationError] if soft == false and validation fails
+      #
       def validate_success_status(soft = true)
```
This method is completely duplicated in multiple places. It would make a lot more sense to move these to a validation module and share the logic.

Lordnibbler replied:
+1

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+        signed_elements = []
+        signature_nodes.each do |signature_node|
+          signed_element = signature_node.parent.name
+          if signed_element != 'Response' and signed_element != 'Assertion'
```
Use `&&` vs. `and` here.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
-            break false
-          else
-            error_message = [error.message, xml.to_s].join("\n\n")
+        unless signature_nodes.length < 3 and not signed_elements.empty?
```
Use `&& !` vs. `and not` here.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
           end
+          valid
+        rescue OneLogin::RubySaml::ValidationError => e
+          @errors << e.message
```
Another useless assignment due to the `raise` below.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -200,27 +487,58 @@ def xpath_first_from_signed_assertion(subelt=nil)
+      #
+      def xpath_from_signed_assertion(subelt=nil)
+        node = REXML::XPath.match(
+            document,
```
This is a more readable style than most other spots touched in this changeset. For consistency, however, please either change this to be one long line, or update the other spots to span multiple lines.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -200,27 +487,58 @@ def xpath_first_from_signed_assertion(subelt=nil)
         now = Time.now.utc
 
-        if not_before && (now + (options[:allowed_clock_drift] || 0)) < not_before
+        if not_before && not_before > (now + (options[:allowed_clock_drift] || 0))
```
It would be nice to see this logic encapsulated in a method like `allowed_clock_drift`, at least the `(options[:allowed_clock_drift] || 0)` part.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -228,20 +546,100 @@ def validate_conditions(soft = true)
+            next
+          else
+            attrs = confirmation_data_node.attributes
+            next if (attrs.include? "InResponseTo" and attrs['InResponseTo'] != in_response_to) ||
```
All of these conditions should use `&&` not `and`.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -228,20 +546,100 @@ def validate_conditions(soft = true)
+            next if (attrs.include? "InResponseTo" and attrs['InResponseTo'] != in_response_to) ||
+                    (attrs.include? "Recipient" and attrs['Recipient'] != current_url) ||
+                    (attrs.include? "NotOnOrAfter" and (parse_time(confirmation_data_node, "NotOnOrAfter") + (options[:allowed_clock_drift] || 0)) <= now) ||
+                    (attrs.include? "NotBefore" and parse_time(confirmation_data_node, "NotBefore") > (now + (options[:allowed_clock_drift] || 0)))
```
So much duplication of `(options[:allowed_clock_drift] || 0)` :\

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -228,20 +546,100 @@ def validate_conditions(soft = true)
+
+          confirmation_data_node = REXML::XPath.first(subject_confirmation, 'a:SubjectConfirmationData', { "a" => ASSERTION })
+
+          if not confirmation_data_node
```
`if !` or `unless`, not `if not`.

phlipper commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -228,20 +546,100 @@ def validate_conditions(soft = true)
+                    (attrs.include? "NotBefore" and parse_time(confirmation_data_node, "NotBefore") > (now + (options[:allowed_clock_drift] || 0)))
+            
+            valid_subject_confirmation = true
+            break
```
What is this `break`ing out of?

pitbulk replied:
If we find a valid_subject_confirmation, then we don't need to loop more the subject_confirmation_nodes

Lordnibbler replied:
I don't see a loop?

pitbulk replied:
https://github.com/onelogin/ruby-saml/pull/197/files#diff-00524fdf114f961e4ff1ff4a04fd8490R602

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -7,27 +7,59 @@
+      # @return [Boolean] TRUE if the Logout Request is valid
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate!(soft = false, get_params = nil)
```
This is another useless wrapper method.

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -50,15 +85,204 @@ def issuer
+          node = REXML::XPath.first(document, "/p:LogoutRequest", { "p" => PROTOCOL} )
+          if node && node.attributes["NotOnOrAfter"]
+            Time.parse(node.attributes["NotOnOrAfter"])
+          else
```
The `else nil` is not needed here.

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -50,15 +85,204 @@ def issuer
-      def validate(soft = true)
-        valid_saml?(document, soft)  && validate_request_state(soft)
+        if nodes
+          nodes.each { |node|
```
inconsistent use of `{ }` vs. `do end`.

Lordnibbler replied:
use `do end` blocks for multi line please

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -50,15 +85,204 @@ def issuer
+
+        # Gets the expected current_url
+        # (Right now we read this url from the Sinle Logout Service of the Settings)
+        # TODO: Calculate the real current_url and use it.
```
:eyes:

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -50,15 +85,204 @@ def issuer
+        # @return [String] The current url
+        #
+        def current_url
+          @current_url ||= begin
```
Memoization not needed here.

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -50,15 +85,204 @@ def issuer
+        # If fails, the error is added to the errors array.
+        # @return [Boolean] True if the Logout Request contains an ID, otherwise returns False
+        #
+        def validate_id()
```
Remove these unneeded parens.

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -50,15 +85,204 @@ def issuer
+        # If fails, the error is added to the errors array.
+        # @return [Boolean] True if the Logout Request is 2.0, otherwise returns False
+        #
+        def validate_version()
```
Remove these unneeded parens.

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -50,15 +85,204 @@ def issuer
+        # @return [Boolean] True if satisfies the conditions, otherwise False if soft=True
+        # @raise [ValidationError] if soft == false and validation fails
+        #
+        def validate_not_on_or_after(soft = true)
```
So much duplication here :`(

phlipper commented on lib/onelogin/ruby-saml/slo_logoutrequest.rb:
```diff
@@ -50,15 +85,204 @@ def issuer
+            end
+            valid
+          rescue OneLogin::RubySaml::ValidationError => e
+            @errors << e.message
```
Another useless assignment due to `raise`.

phlipper commented on lib/onelogin/ruby-saml/slo_logoutresponse.rb:
```diff
@@ -45,11 +67,13 @@ def create_params(settings, request_id = nil, logout_message = nil, params = {})
+            :relay_state => relay_state,
+            :sig_alg     => params['SigAlg']
+          )          
+          signature           = settings.get_sp_key().sign(XMLSecurity::BaseDocument.new.algorithm(settings.security[:signature_method]).new, url_string)
```
Same parens/whitespace/assignment comment as above.

phlipper commented on lib/onelogin/ruby-saml/utils.rb:
```diff
@@ -1,44 +1,89 @@
+      def self.build_query(params)
+        type, data, relay_state, sig_alg = [:type, :data, :relay_state, :sig_alg].map { |k| params[k]}
+
+        url_string          = "#{type}=#{CGI.escape(data)}"
```
Sooo &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;           much &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;          whitespace &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;         here

phlipper commented on ruby-saml.gemspec:
```diff
@@ -1,4 +1,4 @@
-$LOAD_PATH.push File.expand_path('../lib', __FILE__)
+$LOAD_PATH.push File.expand_path("../lib", __FILE__)
```
Inconsistent use of double quotes here and single quotes below.

phlipper commented on test/idp_metadata_parser_test.rb:
```diff
@@ -50,9 +50,37 @@ class MockResponse
+      mock_response = MockResponse.new
+      mock_response.body = idp_metadata_2
+      @url = "http://example.com"
+      uri = URI(@url)
```
Use `URI.parse` to stay consistent with the rest of this changeset.

phlipper commented on test/idp_metadata_parser_test.rb:
```diff
@@ -50,9 +50,37 @@ class MockResponse
+  describe "download and parse IdP descriptor file (HTTP)" do
+    before do
+      mock_response = MockResponse.new
+      mock_response.body = idp_metadata_2
```
Please consider more descriptive names _(or at the very least, use comments)_ for things like `idp_metadata_2`.

phlipper commented on test/idp_metadata_parser_test.rb:
```diff
@@ -50,9 +50,37 @@ class MockResponse
+    before do
+      mock_response = MockResponse.new
+      mock_response.body = idp_metadata_2
+      @url = "http://example.com"
```
Perhaps this would make more sense to use `let`

phlipper commented on test/logoutrequest_test.rb:
```diff
@@ -91,18 +81,24 @@ class RequestTest < Minitest::Test
-        settings.idp_slo_target_url = "http://example.com?field=value"
-        settings.name_identifier_value = "f00f00"
+
+      let(:settings2) { OneLogin::RubySaml::Settings.new }
```
What is `settings2`?

pitbulk replied:
In the test I use 3 different settings: settings, settings2 and settings3, each initiated with different values in the before block ...

In this specific case I think that naming each value with a descriptive name like settings_for_test_signatures, settings_for_test_authcontext .. will do the code less legible due the code length, but I'm open to suggestions.

phlipper replied:
Line length is much less of an issue than non-descriptive names. I can scan to the right, or wrap to the next line, if things get long, As this stands, I need to deal with multiple files to understand the full context of this test.

Lordnibbler replied:
please use more descriptive variable names

phlipper commented on test/logoutresponse_test.rb:
```diff
@@ -30,36 +26,125 @@ class RubySamlTest < Minitest::Test
+        custom_settings = settings
+        custom_settings.idp_entity_id = 'http://app.muda.no'
+
+        logoutresponse = OneLogin::RubySaml::Logoutresponse.new(valid_response({:uuid => in_relation_to_request_id}), custom_settings)
```
This would be more readable if the `valid_response` assignment was on its own line.

phlipper commented on test/logoutresponse_test.rb:
```diff
@@ -30,36 +26,125 @@ class RubySamlTest < Minitest::Test
+        in_relation_to_request_id = random_id
+        bad_settings = settings
+        bad_settings.idp_entity_id = 'http://invalid.issuer.example.com/'
+        logoutresponse = OneLogin::RubySaml::Logoutresponse.new(valid_response({:uuid => in_relation_to_request_id}), bad_settings)
```
Same comment about `valid_response` as above. Also applies to the lines below too.

phlipper commented on test/logoutresponse_test.rb:
```diff
@@ -70,42 +155,160 @@ class RubySamlTest < Minitest::Test
+
+      it "raise validation error for wrong request status and status_message" do
+        unsuccessful_response_with_status_message = unsuccessful_response
+        unsuccessful_response_with_status_message = unsuccessful_response_with_status_message.gsub('</samlp:Status>', '<samlp:StatusMessage>It was requester</samlp:StatusMessage></samlp:Status>')
```
What is the purpose of this `.gsub` operation here?

pitbulk replied:
Instead of creating a new unsuccessful_response example file, I took an existed example and changed it, making it a unsuccessful_response_with_status_message.

phlipper replied:
That is not clear here. Please at least make a comment for that so it is understandable in the future when someone comes back to make changes.

phlipper commented on test/request_test.rb:
```diff
@@ -145,17 +147,48 @@ class RequestTest < Minitest::Test
+    it "create the saml:AuthnContextClassRef element correctly" do
+      settings.authn_context = 'secure/name/password/uri'
+      auth_doc = OneLogin::RubySaml::Authrequest.new.create_authentication_xml_doc(settings)
+      assert auth_doc.to_s =~ /<saml:AuthnContextClassRef>secure\/name\/password\/uri<\/saml:AuthnContextClassRef>/
```
It would be more clear _(and consistent)_ to use `assert_match` for all of these vs. `assert x =~ y`

phlipper commented on test/response_test.rb:
```diff
@@ -34,380 +31,913 @@ class RubySamlTest < Minitest::Test
+      response.settings = settings
+      assert_nil response.name_id
+      response.send(:validate_structure)
+      assert response.errors.include? "Invalid SAML Response. Not match the saml-schema-protocol-2.0.xsd"
```
This can use `assert_includes`

phlipper commented on test/response_test.rb:
```diff
@@ -34,380 +31,913 @@ class RubySamlTest < Minitest::Test
+      response.settings = settings
+      response.settings.idp_entity_id = 'http://idp.example.com/'
+      assert_raises(OneLogin::RubySaml::ValidationError, "Doesn't match the issuer, expected: <#{response.settings.idp_entity_id}>, but was: <http://invalid.issuer.example.com/>") do
+        response.send(:validate_issuer, false)
```
Why is this using `send` vs. simply calling the method? If this is because `validate_issuer` was defined as `private`, perhaps these tests are indicating it should be otherwise.

pitbulk replied:
I could call the public validate method, but other errors can take place instead of the method that I wanted to test. In order to avoid this conflicts on test and make it easier I'm testing each validation method,

phlipper commented on test/utils_test.rb:
```diff
@@ -0,0 +1,217 @@
+    end
+  end
+
+  describe "#build_query" do
```
Either fill these in or remove them.

phlipper commented on test/utils_test.rb:
```diff
@@ -0,0 +1,217 @@
+      cert = cert.gsub('-----BEGIN CERTIFICATE-----', '')
+      cert = cert.gsub('-----END CERTIFICATE-----', '')      
+
+      x509 = OpenSSL::X509::Certificate.new(OneLogin::RubySaml::Utils.format_cert(cert))
```
What is this actually testing? There are no assertions.

phlipper commented on test/xml_security_test.rb:
```diff
@@ -78,7 +78,53 @@ class XmlSecurityTest < Minitest::Test
   end
 
-  describe "Algorithms" do
+  describe "#canon_algorithm" do
```
```ruby
let(:base_document) { XMLSecurity::BaseDocument.new }
# ...
assert_equal canon_algorithm, base_document.canon_algorithm("http://www.w3.org/2001/10/xml-exc-c14n#")
```

phlipper commented:
I appreciate the renewed effort this library is receiving, however I think you need to be very careful about affecting both all of the people currently using it, and the ability to continue maintaining it going forward. There is **a lot** going on here, and as someone who is also working on functionality and changes to this library, PRs like make things harder to understand and harder to change.

Lordnibbler commented:
@phlipper I agree; I'd really like to see this PR broken apart into a few separate PR's, so we can give them the scrutiny they deserve.  @pitbulk @luisvm thoughts?

supernullset commented:
+1 on breaking it apart

pitbulk commented:
I appreciate the review that you did to the code and will fix it asap. As you noticed the gem had some validation required on the SAML standard that were missed. I tested it with IdPs and added code coverage to be sure that all keep working. 

I plan to give time people to test it in master before releasing a new version of the gem.

I apologies for my ruby knowledge. I'm a SAML expert that developed the [php-saml](https://github.com/onelogin/php-saml) and the python-saml[https://github.com/onelogin/php-saml] and doing my best learning ruby fast.

elskwid commented:
@pitbulk I think it's commendable that you are putting so much time and effort into making this a better library. 

If SAMl is your area of expertise, how about a process like this: (1) create tagged issues around features and fixes that are needed, (2) propose those fixes and features in small, concise pull requests, (3) let the community help you with the Ruby aspect of it, (4) merge and be happy. :)

luisvm commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -112,78 +151,326 @@ def status_message
+      def issuers
+        @issuers ||= begin
+          issuers = []
+          nodes = REXML::XPath.match(document, "/p:Response/a:Issuer | /p:Response/a:Assertion/a:Issuer", { "p" => PROTOCOL, "a" => ASSERTION })
```
@pitbulk should we make sure it's the absolute path (`//p:Response/a:Issuer`) or should we keep the relative ones?

Soliah commented:
I'm also in the process of rolling out SAML to one of our products, and started hacking out support for multiple signing certificates from the IdP when I came across this and #195.

So before I just fork and start crazily modifying the project perhaps a discussion around joining efforts would be good?

irvingreid commented on lib/onelogin/ruby-saml/idp_metadata_parser.rb:
```diff
@@ -66,33 +79,65 @@ def get_idp_metadata(url, validate_cert)
+      #
       def idp_entity_id
-        node = REXML::XPath.first(document, "/md:EntityDescriptor/@entityID", { "md" => METADATA })
+        node = REXML::XPath.first(
```
Do all of these value accessors guarantee that the value was fetched from a part of the SAML response that was covered by the XMLDSIG signature? There have been successful attacks on SAML implementations by sending assertions with additional fields that were not covered by the signature, but were trusted because the parts of the assertion that were signed were valid.

pitbulk replied:
Notice that here we are parsing Metadatas, not SAMLResponses

irvingreid replied:
OK.

irvingreid commented on test/response_test.rb:
```diff
@@ -34,380 +31,913 @@ class RubySamlTest < Minitest::Test
-        response.settings = settings
-        assert response.is_valid?
-        assert_equal response.name_id, "test@onelogin.com"
+    it "raise when the destination od the SAML Response does not match the assertion consumer service url" do
```
Typo: s/od/of/

irvingreid commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -89,101 +126,385 @@ def attributes
+        return true if destination.nil? || destination.empty? || settings.assertion_consumer_service_url.nil? || settings.assertion_consumer_service_url.empty?
+
+        unless destination == current_url
+          error_msg = "The response was received at #{destination} instead of #{current_url}"
```
I'd prefer wording something like "Expected destination #{settings.asser...} but the SAML assertion contained #{destination}"

irvingreid commented on lib/onelogin/ruby-saml/response.rb:
```diff
@@ -89,101 +126,385 @@ def attributes
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_destination(soft = true)
+        return true if destination.nil? || destination.empty? || settings.assertion_consumer_service_url.nil? || settings.assertion_consumer_service_url.empty?
```
Spec says that destination must be validated if it comes in the assertion, so validation should fail if settings.assertion_consumer_service_url is nil but the assertion contained a non-nil destination.

pitbulk commented:
We already merge most of the proposed code:
https://github.com/onelogin/ruby-saml/pull/231
https://github.com/onelogin/ruby-saml/pull/232
https://github.com/onelogin/ruby-saml/pull/235
https://github.com/onelogin/ruby-saml/pull/238
https://github.com/onelogin/ruby-saml/pull/239
https://github.com/onelogin/ruby-saml/pull/240
